### PR TITLE
Fix broken compilation in v2.0.0 + additional bugfixes around contexts

### PR DESCRIPTION
**Pull request by tuvalkohl**
### TL;DR
This PR fixes compilation issues and several bugs & race conditions which we introduced were in v2.0.0 (#22).
It also adds more verbosity / comments / explanations around the new `context_info` features.
And also adds more unittest coverage for the `context_info` features.

---
### Explanation about the new context_info features
There are now (as of #22) 3 different levels of `context_info`:
1. `log_context_info` - affects only a single Log
2. `logger_context_info` - affects all logs for a specific Logger
3. `global_context_info`- affects all loggers in the process

* In `JSON` / `CloudWatch` sinks:
If there are conflicting keys between the different contexts, then the most local context takes precedence (i.e order of precedence is the order listed above).

NOTE: The log's `extra_identifier` overrides the `"session_id"` key if it is defined in any of the contexts.

* In `console` / `file` / `syslog` sinks when configured with `PLAINTEXT_LONG`:
If any key appears in more than one context (e.g both in `log_context` and in `logger_context`) then it will be appear multiple times with all the different values. 

---
### Main fix in this PR:
**[thread-safety] IMPORTANT:** Wrap all access to the `global_context_info_` shared pointer in `atomic store/load` functions to ensure that the shared global_context_info is accessed in a thread-safe manner.

The actual ContextInfo that we point at is immutable (const), so there's no risk
in accessing it from multiple threads.
However, it is *not* thread-safe to have concurrent access to the *pointer address*
held in the global_context_info_ shared pointer, since the address itself might be
modified by a different thread if replace/update global_context_info is called.

Theoretically (to my understanding), if we copy/access the pointer held
by the shared_ptr while it is being concurrently replaced by another
thread, we might end up working on a corrupted address. Using atomic
store/load operations protects from this risk.

See:
https://stackoverflow.com/questions/40223599/what-is-the-difference-between-stdshared-ptr-and-stdatomicstdshared-ptr
https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic
https://ddanilov.me/usage-of-shared_ptr

---
### Additional changes/fixes in this PR:

1. bugfix: `update_global_context_info` now updates properly so that overridden keys will actually override

Example:
```
auto old_ci = manager.global_context_info();
ContextInfo ci{{"key1", "value1"}, {"key2", "value2"}, {"key3", "value3"}};
manager.replace_global_context_info(ci);
manager.update_global_context_info({{"key2", "new_value2"}});
```
The above code would result in `"key2"` having `"value2"` instead of being overwritten with `"new_value_2"`.

2. Add a `replace_global_context_info_rvalue` method.
Note: It wasn't possible to overload `replace_global_context_info` because it resulted in ambiguity for the compiler (since `ContextInfo` also has a ctor which receives an rvalue)

3. Add `ContextInfo::at` method 
4. Add a unittest for `update_global_context_info` overriding an existing key
5. Bump version `2.0.0` -> `2.0.1`
6. JSON formatted timestamp will now include milliseconds (still ISO 8601 compliant) - `YYYY-MM-DDTHH:MM:SS.mmm±HHMM`
7. `global_context_info_` now initialized with `make_shared` instead of `make_unique` (I think the latter was leftover by mistake)
8. `Manager::global_context_info()` now returns a `shared_ptr` instead of the pointed-at Context object, which might otherwise be released while being used.
9. `ContextInfo` dtor made non-virtual, since there is no inheritance from this class (and I don't expect any either)
10. Switch to using structured bindings which is nicer than using `itr.first`/`itr.second`
11. Added tests for all the new log/logger/global context_info functionality in the `cloudwatch-sink-tests`

### CODE REVIEW

**Comment by tuvalkohl** [comment]
Need to check failed unittests
```
-------------------------------------------------------------------------------
/Users/tuval.kohl/work/adb-ws/3rdparties/adb-octo-logger-cpp/source/unittests/src/cloudwatch-sink-tests.cpp:151
...............................................................................

/Users/tuval.kohl/work/adb-ws/3rdparties/adb-octo-logger-cpp/source/unittests/src/cloudwatch-sink-tests.cpp:249: FAILED:
  REQUIRE_THAT( itr_context_info.dst, JSONEquals(itr_context_info.expected_result) )
with expansion:
  {"property_1":"property 1 value","property_2":"property 2 value",
  "session_id":"\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\
  u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\
  u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\
  u0000\u0000"}
  equals
  {"property_1":"property 1 value","property_2":"property 2 value",
  "session_id":"69426e8b-83a8-47f1-9e09-87857819db75"}
with messages:
  itr_context_info.session_id := ""
  itr_context_info.dst := {"property_1":"property 1 value"}
  itr_context_info.context_info := { {?}, {?} }

===============================================================================
test cases:   7 |   6 passed | 1 failed
assertions: 137 | 136 passed | 1 failed

**Review by ofiriluz — APPROVED** [review]

File: include/octo-logger-cpp/context-info.hpp [hunk 1]
```diff
@@ -32,7 +32,7 @@ class ContextInfo
     ContextInfoType context_info_;
 
   public:
-    virtual ~ContextInfo() = default;
+    ~ContextInfo() = default;
```
Comment (aradyaron) [inline include/octo-logger-cpp/context-info.hpp:35]: requested by Denis before

**Review by aradyaron — CHANGES_REQUESTED** [review]
All looks good to me, please add a test with a check of global context info usage and Approved.
also, please verify the milliseconds format

File: include/octo-logger-cpp/context-info.hpp (hunk 1 above)
Comment (tuvalkohl) [reply include/octo-logger-cpp/context-info.hpp:35]: Spoke to Denis and marked the class as `final`

File: src/manager.cpp [hunk 2]
```diff
@@ -192,20 +198,32 @@ bool Manager::mute_channel(std::string const& name)
...
+{
+    // First allocate the new ContextInfo, and then atomically replace the pointer held in global_context_info_
+    auto new_context_info = std::make_shared<ContextInfo const>(context_info);
+    std::atomic_store(&global_context_info_, new_context_info);
 }
 
-void Manager::update_global_context_info(ContextInfo context_info)
+// Calling this method concurrently from multiple threads could result in loss of context info (one of the calls could
+// be lost)
+void Manager::update_global_context_info(ContextInfo const& new_context_info)
```
[the remark this answers is not in the document]
Comment (tuvalkohl) [reply src/manager.cpp:220]: Not sure I understand your question.
I fixed here a bug in the implementation, previously updating wouldn't actually update existing kets.

Correct me if I'm wrong, but I don't think the performance of `update_global_context_info`is super-critical, since it's not expected to be called often, only upon specific events which change the context.

See what I wrote in the PR description:
1. bugfix: `update_global_context_info` now updates properly so that overridden keys will actually override

Example:
```
auto old_ci = manager.global_context_info();
ContextInfo ci{{"key1", "value1"}, {"key2", "value2"}, {"key3", "value3"}};
manager.replace_global_context_info(ci);
manager.update_global_context_info({{"key2", "new_value2"}});
```
The above code would result in `"key2"` having `"value2"` instead of being overwritten with `"new_value_2"`.

**Comment by tuvalkohl** [comment]
> All looks good to me, please add a test with a check of global context info usage and Approved. also, please verify the milliseconds format

Added tests for log/logger/global context_info in the cloudwatch-sink-tests...

Re milliseconds format, I tested this code locally and confirmed that the output is ISO 8601 compliant.

**Review by aradyaron — APPROVED** [review]
